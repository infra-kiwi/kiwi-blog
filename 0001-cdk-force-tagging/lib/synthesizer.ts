import * as cdk from 'aws-cdk-lib';
import simpleGit from 'simple-git';

export type KiwiSynthesizerProps = cdk.DefaultStackSynthesizerProps & {
  forcedTags: Record<string, string>;
};

export class KiwiSynthesizer extends cdk.DefaultStackSynthesizer {
  readonly forcedTags: KiwiSynthesizerProps['forcedTags'];

  constructor(props: KiwiSynthesizerProps) {
    super(props);
    this.forcedTags = props.forcedTags;
  }

  bind(stack: cdk.Stack) {
    super.bind(stack);

    // Add every forced tag to the Stack
    for (const tag in this.forcedTags) {
      cdk.Tags.of(this.boundStack).add(tag, this.forcedTags[tag]);
    }
  }
}

export async function getSynthesizer(synthProps?: cdk.DefaultStackSynthesizerProps) {
  // Extract the current repository name
  const remoteUrl = (await simpleGit().remote(['get-url', 'origin']))!.trim();
  const repository = /github.com[:/]([^/]+\/[^/]+)\.git/.exec(remoteUrl)![1];

  return new KiwiSynthesizer({
    ...synthProps,
    forcedTags: {
      Repository: repository
    }
  });
}
