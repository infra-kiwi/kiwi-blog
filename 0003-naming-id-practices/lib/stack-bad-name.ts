/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NotificationGroup } from './constructs-bad-name/notification-group';
import { SNSTopicWrapper } from './constructs-bad-name/sns-topic-wrapper';

export class StackBadName extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new SNSTopicWrapper(this, 'NotificationsFunFacts');

    new NotificationGroup(this, 'Default');
    new NotificationGroup(this, 'TeamInfra');
  }
}
