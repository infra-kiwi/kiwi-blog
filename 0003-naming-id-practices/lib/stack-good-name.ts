import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NotificationGroup } from './constructs-good-name/notification-group';
import { SNSTopicWrapper } from './constructs-good-name/sns-topic-wrapper';

export class StackGoodName extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new SNSTopicWrapper(this, 'NotificationsFunFacts');

    new NotificationGroup(this, 'Default');
    new NotificationGroup(this, 'TeamInfra');
  }
}
