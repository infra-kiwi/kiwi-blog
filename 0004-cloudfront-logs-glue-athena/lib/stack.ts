import * as cdk from 'aws-cdk-lib';
import { aws_cloudfront, aws_s3, CfnOutput, RemovalPolicy, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import * as aws_glue_alpha from '@aws-cdk/aws-glue-alpha';

export class CloudFrontStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a dedicated S3 bucket to store CloudFront logs
    const logBucket = new aws_s3.Bucket(this, 'CloudFrontLogs', {
      // ACL, required by CloudFront
      objectOwnership: aws_s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
      // Make sure we don't keep unnecessary data after we delete this stack
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true
    });

    const distribution = new aws_cloudfront.Distribution(this, 'Distribution', {
      // HTTPS (because we should ALWAYS use HTTPS!)
      sslSupportMethod: aws_cloudfront.SSLMethod.SNI,

      // Enable logging to the S3 bucket
      enableLogging: true,
      logBucket: logBucket,

      // Use the wonderful https://reqres.in/ as a destination for our calls
      defaultBehavior: {
        origin: new HttpOrigin('reqres.in')
      }
    });
    // This output will contain the domain name of the distribution,
    // e.g. xxxxxxx.cloudfront.net
    new CfnOutput(this, 'DistributionDomainName', {
      value: distribution.domainName
    });

    // Create a database where to store the table
    const database = new aws_glue_alpha.Database(this, 'GlueDatabase', {
      // Give the database a friendly name (lower-case only!)
      databaseName: Stack.of(this).stackName + '-database'
    });
    new CfnOutput(this, 'CloudFrontLogsDatabaseName', {
      value: database.databaseName
    });

    // Create the table that maps each log line
    const table = new aws_glue_alpha.S3Table(this, 'CloudFrontLogsTable', {
      // Give the table a friendly name (lower-case only!)
      tableName: Stack.of(this).stackName + '-cloudfrontlogstable',

      // The source bucket for this Glue table
      bucket: logBucket,
      database: database,

      // Full SQL reference: https://docs.aws.amazon.com/athena/latest/ug/cloudfront-logs.html
      columns: [
        { name: 'date', type: aws_glue_alpha.Schema.DATE },
        { name: 'time', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_location', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_bytes', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'c_ip', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_method', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_host', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_uri_stem', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_status', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'cs_referrer', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_user_agent', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_uri_query', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_cookie', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_request_id', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_host_header', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_protocol', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_bytes', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'time_taken', type: aws_glue_alpha.Schema.FLOAT },
        { name: 'x_forwarded_for', type: aws_glue_alpha.Schema.STRING },
        { name: 'ssl_protocol', type: aws_glue_alpha.Schema.STRING },
        { name: 'ssl_cipher', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_response_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_protocol_version', type: aws_glue_alpha.Schema.STRING },
        { name: 'fle_status', type: aws_glue_alpha.Schema.STRING },
        { name: 'fle_encrypted_fields', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'c_port', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'time_to_first_byte', type: aws_glue_alpha.Schema.FLOAT },
        { name: 'x_edge_detailed_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_content_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_content_len', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'sc_range_start', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'sc_range_end', type: aws_glue_alpha.Schema.BIG_INT }
      ],

      // Define the right settings to parse CloudFront logs
      dataFormat: aws_glue_alpha.DataFormat.TSV,
      parameters: {
        // THESE PARAMETERS ARE VERY IMPORTANT
        'skip.header.line.count': '2',
        'serialization.format': '\t',
        'field.delim': '\t'
      }
    });
    new CfnOutput(this, 'CloudFrontLogsTableName', {
      value: table.tableName
    });
  }
}
