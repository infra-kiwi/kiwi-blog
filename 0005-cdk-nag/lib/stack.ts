/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_kms, aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Use the default AWS-provided SNS KMS key
    const awsSNSKey = aws_kms.Alias.fromAliasName(
      this,
      'AWSManagedSNSKey',
      'alias/aws/sns'
    );

    this.topic = new aws_sns.Topic(this, 'MyTopic', {
      masterKey: awsSNSKey
    });
    new CfnOutput(this, 'MyTopicArn', {
      value: this.topic.topicArn
    });
  }
}
