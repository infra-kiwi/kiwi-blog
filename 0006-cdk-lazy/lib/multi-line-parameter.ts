/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import { Construct } from 'constructs';
import { aws_ssm, CfnOutput, Lazy } from 'aws-cdk-lib';

export interface MultiLineParameterProps {
  parameterName: string;
}

export class MultiLineParameter extends Construct {
  readonly parameter: aws_ssm.StringParameter;

  // Stores temporarily the lines of our parameter's text
  private readonly lines: string[] = [];

  constructor(scope: Construct, id: string, props: MultiLineParameterProps) {
    super(scope, `${MultiLineParameter.name}-${id}`);

    // Define the wrapped parameter
    this.parameter = new aws_ssm.StringParameter(this, 'Parameter', {
      parameterName: props.parameterName,

      /*
      Here we make use of the Lazy function to assemble
      the lines of the text only at synth time.
       */
      stringValue: Lazy.string({ produce: () => this.lines.join('\n') })
    });

    // Print the name of the parameter we created
    new CfnOutput(this, 'ParameterName', {
      value: this.parameter.parameterName
    });
  }

  /**
   * `addLine` adds a line of text to the parameter
   *
   * @param line
   */
  addLine(line: string) {
    this.lines.push(line);
  }
}
