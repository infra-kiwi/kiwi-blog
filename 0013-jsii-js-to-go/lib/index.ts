/*
 * Copyright (c) 2025. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

export class Constants {
  static readonly name = 'Mario';
}

export interface Home {
  readonly address: string;
}

export class Addressor {
  getFullGreeting(name: string, home: Home) {
    return `Hello ${Constants.name}! You live at ${home.address}.`;
  }
}
