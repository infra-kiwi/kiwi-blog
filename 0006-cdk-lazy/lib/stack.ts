import * as cdk from 'aws-cdk-lib';
import { aws_sns, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MultiLineParameter } from './multi-line-parameter';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const myParameter = new MultiLineParameter(this, 'MyParameter', {
      parameterName: Stack.of(this).stackName + '-my-parameter'
    });

    // Populate the lines of our multi-line parameter
    myParameter.addLine("It's not DNS");
    myParameter.addLine("There's no way it's DNS");
    myParameter.addLine('It was DNS');

    // Add a dynamic line that depends on another CDK resource's value
    const aTopicAboutDNS = new aws_sns.Topic(this, 'ATopicAboutDNS');
    myParameter.addLine(
      `Here is a topic about DNS: ${aTopicAboutDNS.topicArn}`
    );
    this.topic = aTopicAboutDNS;
  }
}
