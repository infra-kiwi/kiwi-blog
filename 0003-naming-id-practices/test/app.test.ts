/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { StackGoodName } from '../lib/stack-good-name';
import { StackBadName } from '../lib/stack-bad-name';

test('Naming stacks', async () => {
  const app = new App();

  // Init stacks
  const stackGoodName = new StackGoodName(app, 'GoodName');
  const stackBadName = new StackBadName(app, 'BadName');

  // Prepare for assertions
  {
    const template = Template.fromStack(stackGoodName);
    expect(template).toMatchSnapshot();

    template.resourceCountIs('AWS::SNS::Topic', 5);
  }

  {
    const template = Template.fromStack(stackBadName);
    expect(template).toMatchSnapshot();

    template.resourceCountIs('AWS::SNS::Topic', 5);
  }
});
