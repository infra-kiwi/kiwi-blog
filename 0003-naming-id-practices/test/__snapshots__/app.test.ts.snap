// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Naming stacks 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "NotificationGroupDefaultSNSTopicWrapperInfraNotificationsTopic7482A443": {
      "Type": "AWS::SNS::Topic",
    },
    "NotificationGroupDefaultSNSTopicWrapperSlackMessagesTopic564777CA": {
      "Type": "AWS::SNS::Topic",
    },
    "NotificationGroupTeamInfraSNSTopicWrapperInfraNotificationsTopic33A0C436": {
      "Type": "AWS::SNS::Topic",
    },
    "NotificationGroupTeamInfraSNSTopicWrapperSlackMessagesTopic9EA8DCBC": {
      "Type": "AWS::SNS::Topic",
    },
    "SNSTopicWrapperNotificationsFunFactsTopic9BB85BA8": {
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Naming stacks 2`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "InfraNotificationsTopic55016246": {
      "Type": "AWS::SNS::Topic",
    },
    "NotificationsFunFactsTopicF7A881FD": {
      "Type": "AWS::SNS::Topic",
    },
    "SlackMessagesTopic0D4141FC": {
      "Type": "AWS::SNS::Topic",
    },
    "TeamInfraInfraNotificationsTopicA59C0E27": {
      "Type": "AWS::SNS::Topic",
    },
    "TeamInfraSlackMessagesTopic5C3FC0A6": {
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
